ENH: support migration, validation, schema improvements

satra opened the pull request:
These schema updates are being used both to fix missing elements in the schema (like schemaKey) but also to handle asset metadata being discussed in this PR: https://github.com/dandi/dandi-cli/pull/644

- Fix: missing schemaKey and autogenerate them
- Enh: Migration from older schema
- Enh: Validation of dandiset
- RF: Move datacite support out of models
- RF: move datacite terms into their own namespace
- FIX: relatedResource now supports doi's and URLs
- TST: improve test coverage
- RF: JSON data
- RF: make pydantic validation primary

satra commented:
@yarikoptic - Equipment is added as an entity that was used in a Session rather than as an Agent. This will get more complicated with self-driving equipment, but for the moment most equipment are passive and therefore have no agency.

satra commented:
@yarikoptic - this should be good to go for now. how do i create a release?

yarikoptic commented:
If all works, just add release and minor or major or patch label and click merge. I added labels, you click merge? ;-)

yarikoptic commented:
The question though - do we want to release or stage more changes into it?

satra commented:
> do we want to release or stage more changes into it? 

this will help use it in the API server. also this schema change will require schema version updates on both the UI and the API. 

we will likely need to do a mini-migration for some dandiset metadata. actually, let me add a migrate function to this. probably will need refinement over time.

yarikoptic commented:
I wouldn't bother with migration ATM. We still should be able to reextract and upload (since they don't mutate ATM ;-))

satra commented:
realized that i wasn't paying attention to the dandi namespace, so had conflicts between roletypes and top level items (Software, EthicsApprovals). also changed Open to OpenAccess and added a dandiset metadata migration function.

for assets we would mint new metadata, but dandiset metadata also needs to be migrated alongside schema change on the server. so after this is merged, we do need to do the updates simultaneously.

wouldn't mind some eyes looking at the changes before we merge.

yarikoptic commented on dandischema/model_types.py:
```diff
@@ -422,169 +422,169 @@
         },
         {
-            "@id": "dandi:Author",
+            "@id": "dandirole:Author",
```
so it is `dandi:RestrictedAccess` (thus `Access` is folded into value) but `dandirole:Author` and not e.g. `dandi:AuthorRole` ... ideally should be consistent

satra replied:
let me think about this. it is indeed our namespace so we can control.

satra replied:
i moved everything for datacite to a `dcite` prefix. unfortunately they don't have a url so i just used a dandi namespace for it.

yarikoptic commented on dandischema/tests/test_datacite.py:
```diff
@@ -0,0 +1,277 @@
+
+
+def datacite_post(datacite, doi):
+    """ posting the datacite object and checking the status of the requests"""
```
```suggestion
    """Post the datacite object and check the status of the request"""
```
I think imperative form is generally better and that is what we use primarily: it is shorter, `-ing` suggests some extended in time action (content manager? generator? nope...), and I do not see a point in leading space.

yarikoptic commented on dandischema/tests/test_datacite.py:
```diff
@@ -0,0 +1,277 @@
+
+
+def _clean_doi(doi):
+    """removing doi, ignoring the status code"""
```
```suggestion
    """Remove doi. Status code is ignored"""
```

yarikoptic commented on dandischema/tests/test_datacite.py:
```diff
@@ -0,0 +1,277 @@
+
+
+def _basic_publishmeta(dandi_id, version="v.0", prefix="10.80507"):
+    """
+    adding basic info required by PublishedDandisetMeta in addition to
+    fields required by DandisetMeta
+    """
```
```suggestion
    """Return extra metadata required by PublishedDandisetMeta 
    
    Returned fields are additional to fields required by DandisetMeta
    """
```

yarikoptic commented on dandischema/metadata.py:
```diff
@@ -90,3 +95,30 @@ def validate_asset_json(data, schema_dir):
+    klass(**obj)
+
+
+def migrate_001(obj):
```
also how is this one interfaced, or nohow yet? altogether, ideally, I would have kept it for another PR which would have e.g. added support for migrations and allow to validate older schemas upon automagic migration.

yarikoptic commented on dandischema/tests/test_metadata.py:
```diff
@@ -0,0 +1,238 @@
+def test_asset(schema_dir):
+    with (METADATA_DIR / "asset_001.json").open() as fp:
+        data_as_dict = json.load(fp)
+    data_as_dict["schemaVersion"] = DANDI_SCHEMA_VERSION
```
with migrations added, I think such overwrites of schemaVersion should not be needed and not desired.  But I guess for now it is ok, but let's make that explicit

```suggestion
    # overload (here and below) schemaVersion until we support automagic schema migrations
    # under assumption that the 0.3.2 schema would be forward compatible
    data_as_dict["schemaVersion"] = DANDI_SCHEMA_VERSION
```

yarikoptic reviewed (commented):
Well, can't make verdict one way or another ATM, but suggested some changes and asked about clarification for `dandi:` -> `dandirole:` necessity .

Migrations -- might need a more thorough thinking through on how/if we could support those.

satra commented on dandischema/metadata.py:
```diff
@@ -3,17 +3,19 @@
+    field_preamble = {
         "@version": 1.1,
         "dandi": "http://schema.dandiarchive.org/",
+        "dandirole": "http://schema.dandiarchive.org/roles/",
```
to make the uris address different concepts. this is the same thing as: `dandi:roles/Software` but i was using the prefix to parse certain things.

satra commented on dandischema/metadata.py:
```diff
@@ -90,3 +95,30 @@ def validate_asset_json(data, schema_dir):
+
+def migrate_001(obj):
+    schema_version = obj.get("schemaVersion")
+    if schema_version == DANDI_SCHEMA_VERSION:
```
this should work as `DANDI_SCHEMA_VERSION` is always the latest. so if something conforms to that, don't make any changes. it's a trivial check, but i think it does the right thing.

jwodder commented on dandischema/metadata.py:
```diff
@@ -90,3 +95,50 @@ def validate_asset_json(data, schema_dir):
     jsonschema.validate(data, schema)
+
+
+def validate(obj, schema_version=None, schema_key=None):
```
This docstring isn't terminated.

satra replied:
yup - will be fixed in the update. got sidetracked on improving models.

satra commented on dandischema/metadata.py:
```diff
@@ -90,3 +95,30 @@ def validate_asset_json(data, schema_dir):
+    if schema_key == "Dandiset":
+        schema_key = "DandisetMeta"
+    elif schema_key == "Asset":
+        schema_key = "AssetMeta"
```
done

satra commented:
@yarikoptic - this should be ready now.

satra commented:
@jwodder, @yarikoptic  - i created a new access token for our bot user for this repo and set the secret with it.

yarikoptic commented:
lost in comments:  I did add `release` label awhile back, but then I thought I saw a comment from you @satra that we might not want to release just right away so I disabled releasing schemata with #11 in favor of @jwodder working out a proper release support according to #2 .  So -- what is your desire -- to have it released asap upon merge?

yarikoptic commented on .github/workflows/test.yml:
```diff
@@ -3,12 +3,13 @@ name: Tests
   push:
   pull_request:
-  schedule:
-    - cron: '0 6 * * *'
```
why?   It is good to have things tested regularly to make sure that it still all works eg. when new pydantic comes out etc.
 
I haven't spotted any commit message summary among 22 (so didn't go one by one) to see if more details would be given.

NB filing as a separate comment to not await for my full review

satra replied:
it seems to be solving a different problem. it would be good to use less electrons when possible. once we get into lazy mode with schema, we can turn cron job back on. or if we did this once a week or month. every 6 hours is total overkill.

jwodder replied:
That's not "every 6 hours"; it's "once a day at 6 AM".

satra replied:
that is true.

yarikoptic replied:
pushed bringing back daily run - the change was OOS for this PR ;)

satra commented:
yes this releases schema 0.4.0 so should happen on release of this library. so merge -> release library -> release schema. the last two go hand in hand.

yarikoptic commented on dandischema/models.py:
```diff
@@ -1088,16 +905,18 @@ class BareAssetMeta(CommonModel):
     )
 
-    path: str = Field(None, nskey="dandi")
-
```
is order of specification here reflects itself in metadata records dumps?  then we might as well harmonize as to place all mandatory first (may be sort them alphabetically) and then all optional next, also alphabetically.

Special cases -- read-only identifier first?  the rest of read-only fields last.

Might be feasible to do automagically in that metaclass I proposed above. if done prior feeding the list of attributes to super class (so then `schemaKey` would be last, as it should ;)?)

satra replied:
i wouldn't rely on ordering here i just moved it to be visually closer to mandatory fields

yarikoptic commented on dandischema/models.py:
```diff
@@ -1133,42 +952,57 @@ class BareAssetMeta(CommonModel):
 
+    @validator("digest")
+    def digest_etag(cls, values):
+        print(values)
```
```suggestion
```
I guess a left over

yarikoptic commented on dandischema/models.py:
```diff
@@ -1133,42 +952,57 @@ class BareAssetMeta(CommonModel):
+            if len(values[DigestType.dandi_etag]) != 32:
+                raise ValueError
+        except (KeyError, ValueError):
+            raise ValueError("Digest must have an appropriate dandi-etag value.")
```
```suggestion
            raise ValueError(f"Digest must have an appropriate dandi-etag value. Got {values[DigestType.dandi_etag]}")
```
what can I say -- I am a friend of mortals ;-) (note: might need linting)

I assume that the `@validator` would wrap those nicely into `pydantic.ValidationError`s?

yarikoptic commented on dandischema/models.py:
```diff
@@ -1133,42 +952,57 @@ class BareAssetMeta(CommonModel):
+            if len(values[DigestType.dandi_etag] + values[DigestType.sha2_256]) != 96:
+                raise ValueError
+        except (KeyError, ValueError):
+            raise ValueError("Digest must have both dandi-etag and sha2-256.")
```
```suggestion
            raise ValueError("Digest must have both valid dandi-etag and sha2-256.")
```
I would have loved to stick actual values but decided not to bother yet - would need assignment to temp vars above

yarikoptic commented on dandischema/models.py:
```diff
@@ -315,6 +96,7 @@ def encode(self, o):
 
 class DandiBaseModel(BaseModel):
     id: Optional[str] = Field(description="Uniform resource identifier", readOnly=True)
+    schemaKey: Literal["DandiBaseModel"] = Field("DandiBaseModel", readOnly=True)
```
I will submit this review now and quickly try to furnish a PR against this PR branch to introduce such metaclass to see if feasible.

yarikoptic reviewed (changes requested):
- minor suggestions or questions (e.g. re `Type` classes) here and there to strengthen code and unify naming
- `metaclass` idea: to avoid manual schemaKey entry causing duplication etc.  Will furnish a PR quickly to see if it is worth it
- etc ;-)

satra commented on dandischema/datacite.py:
```diff
@@ -0,0 +1,190 @@
+            if len(contr_el.name.split(",")) == 2:
+                contr_dict["familyName"], contr_dict["givenName"] = contr_el.name.split(
+                    ","
+                )
```
i can't remember if i put a regex in the name field for person, but will check.

yarikoptic replied:
filed dedicated https://github.com/dandi/dandischema/issues/16

satra commented on dandischema/datacite.py:
```diff
@@ -0,0 +1,190 @@
+            if contr_all:
+                contr_dict["contributorType"] = contr_all[0]
+            else:
+                contr_dict["contributorType"] = "Other"
```
there are many things to do, also not all roles we have are in datacite yet. they will eventually get there, but not yet. so i don't want to cripple our system to work just for datacite.

satra commented on dandischema/datacite.py:
```diff
@@ -0,0 +1,190 @@
+    if getattr(meta, "relatedResource"):
+        attributes["relatedIdentifiers"] = []
+        for rel_el in meta.relatedResource:
+            ident = rel_el.identifier.split(":")
```
this is true. that split won't work for the pattern we have.

satra commented on dandischema/metadata.py:
```diff
@@ -3,17 +3,20 @@
+    field_preamble = {
         "@version": 1.1,
         "dandi": "http://schema.dandiarchive.org/",
+        "dcite": "http://schema.dandiarchive.org/datacite/",
```
thank you.

satra commented on dandischema/metadata.py:
```diff
@@ -3,17 +3,20 @@
         "@version": 1.1,
         "dandi": "http://schema.dandiarchive.org/",
+        "dcite": "http://schema.dandiarchive.org/datacite/",
         "dandiasset": "http://iri.dandiarchive.org/",
```
yes, the schema.dandiarchive.org are for terms/classes/properties not identifiers. i will change to this last form.

satra commented on dandischema/metadata.py:
```diff
@@ -61,32 +65,89 @@ def generate_context():
 
 
-def validate_asset_json(data, schema_dir):
+def validate_asset_json(data: dict, schema_dir: str) -> None:
```
```suggestion
def _validate_asset_json(data: dict, schema_dir: str) -> None:
```

satra commented on dandischema/metadata.py:
```diff
@@ -61,32 +65,89 @@ def generate_context():
+    if to_version not in ALLOWED_TARGET_SCHEMAS:
+        raise ValueError(f"Current target schemas: {ALLOWED_TARGET_SCHEMAS}.")
+    schema_version = obj.get("schemaVersion")
+    if schema_version == DANDI_SCHEMA_VERSION:
```
yes

yarikoptic replied:
filed dedicated https://github.com/dandi/dandischema/issues/14 so we do not forget

satra commented on dandischema/metadata.py:
```diff
@@ -61,32 +65,89 @@ def generate_context():
+        for access in obj["access"]:
+            access["status"] = "dandi:OpenAccess"
+    obj["schemaVersion"] = to_version
+    return obj
```
because validation may still fail. the only time validation passes will be when required fields are in place. migrate can be called on non-publishable metadata.

satra commented on dandischema/models.py:
```diff
@@ -1052,13 +869,13 @@ def check_data(cls, values):
 
     # From server (requested by users even for drafts)
-    manifestLocation: TempOptional[List[HttpUrl]] = Field(readOnly=True, nskey="dandi")
+    manifestLocation: List[HttpUrl] = Field(readOnly=True, nskey="dandi")
```
metadata saved does not have to be valid until publish. however, while `assetsSummary` can change and should be updated based on asset related triggers, `manifestLocation` does not change once a dandiset is created and/or published. so it will be considered invalid till those fields are populated. but that's a state a dandiset metadata can be in without much effort (for example no ContactPerson entered or removed).

yarikoptic commented on dandischema/models.py:
```diff
@@ -380,10 +176,13 @@ def schema_extra(schema: Dict[str, Any], model) -> None:
                 if prop == "schemaKey":
-                    if len(value["enum"]) == 1:
+                    if "enum" in value and len(value["enum"]) == 1:
                         value["const"] = value["enum"][0]
```
absent coverage here confirms that it doesn't happen across CIs either after moving to custom metaclass, hm... good or bad? ;)

yarikoptic commented on dandischema/models.py:
```diff
@@ -425,47 +208,47 @@ class TypeModel(DandiBaseModel):
 
 
-class Disorder(TypeModel):
+class Disorder(BaseType):
```
filed dedicated https://github.com/dandi/dandischema/issues/15

yarikoptic commented on dandischema/datacite.py:
```diff
@@ -0,0 +1,192 @@
+    ]
+    attributes["publisher"] = "DANDI Archive"
+    attributes["publicationYear"] = str(meta.datePublished.year)
+    # not sure about it dandi-api had "resourceTypeGeneral": "NWB"
```
note: lost track of comments... I was wondering there about where we could get specific  data standards -- we would have them in `AssetsSummary. dataStandard: Optional[List[StandardsType]] `, but not sure where we would stick them yet in datacite

yarikoptic reviewed (approved):
To not hold progress -- filed dedicated issues on comments which weren't addressed yet as part of this PR. I think we would be in greater shape with this magnificent change (thanks @satra for doing the exciting metadata schema stuff!) so let's proceed. Personally I would have preferred to massage it more to get "well done" given the number of components reliant on schema, but of cause we can iterate.

satra commented:
closes #14, #16
